File: ExpressionPL/src/main/java/br/unb/cic/epl/features/add/Add.java
package br.unb.cic.epl.features.add;

import br.unb.cic.epl.base.Exp;

public class Add implements Exp {

    private Exp left;
    private Exp right;

    public Add(Exp left, Exp right) {
	this.left = left;
	this.right = right;
    }

    public String print() {
	return left.print() + " + " + right.print();
    }

}

File: ExpressionPL-scala/src/main/scala/br/unb/cic/epl/Add.scala
package br.unb.cic.epl


package object Add {

  trait AbstractAdd extends Core.Expression {
    type T <: Core.Expression

    // lhs and rhs sao do tipo T, que tem como
    // base o tipo Core.Expression
    var lhs: T = _
    var rhs: T = _

    override def print(): String = "(" + lhs.print() + "+" + rhs.print() + ")" 
  }

  class Add(l: Core.Expression, r: Core.Expression) extends AbstractAdd {
    type T = Core.Expression

    lhs = l
    rhs = r
  }

}
